README.md

# GitHub Label Sync [![NPM version](https://img.shields.io/npm/v/github-label-sync.svg)](https://www.npmjs.com/package/github-label-sync) [![Build status](https://img.shields.io/circleci/project/Financial-Times/github-label-sync.svg)](https://circleci.com/gh/Financial-Times/github-label-sync) [![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)][license]

Synchronise your GitHub labels with as few destructive operations as possible – similar labels get renamed.

Table Of Contents
-----------------

- [Requirements](#requirements)
- [Command-Line Interface](#command-line-interface)
- [JavaScript Interface](#javascript-interface)
- [Label Config File](#label-config-file)
- [Configuration](#configuration)
- [Contributing](#contributing)
- [License](#license)


Requirements
------------

You'll need [Node.js][node] 12+ installed to run GitHub Label Sync. You'll also need a GitHub access token ready so that the the tool will have access to your repositories. You can [generate an access token here][access-tokens], be sure to allow the "repo" scope.


Command-Line Interface
----------------------

Install GitHub Label Sync globally with [npm][npm]:

```sh
npm install -g github-label-sync
```

This installs the `github-label-sync` command-line tool:

```
Usage: github-label-sync [options] <repository>

Options:

  -h, --help                  output usage information
  -V, --version               output the version number
  -a, --access-token <token>  a GitHub access token (also settable with a GITHUB_ACCESS_TOKEN environment variable)
  -l, --labels <path>         the path or URL to look for the label configuration in. Default: labels.json
  -d, --dry-run               calculate the required label changes but do not apply them
  -A, --allow-added-labels    allow additional labels in the repo, and don't delete them
```

Run GitHub Label Sync on a repo (reading [label data](#label-config-file) from a local `labels.json`):

```sh
github-label-sync --access-token xxxxxx myname/myrepo
```

Run GitHub Label Sync using a different [label config file](#label-config-file):

```sh
github-label-sync --access-token xxxxxx --labels my-labels.json myname/myrepo
```

[Label config file](#label-config-file) can be also specified as YAML:

```sh
github-label-sync --access-token xxxxxx --labels my-labels.yml myname/myrepo
```

Perform a dry run, only making safe "read" requests to the GitHub API:

```sh
github-label-sync --access-token xxxxxx --dry-run myname/myrepo
```

Normally any additional labels found on the repo are deleted. Run GitHub label sync and _ignore_ additional labels, leaving them as-is:

```sh
github-label-sync --access-token xxxxxx --allow-added-labels myname/myrepo
```


JavaScript Interface
--------------------

Install GitHub Label Sync with [npm][npm] or add to your `package.json`:

```
npm install github-label-sync
```

Require GitHub Label Sync:

```js
var githubLabelSync = require('github-label-sync');
```

The `githubLabelSync` function returns a promise that resolves to a JSON diff between the labels found on GitHub, and the labels in your label config.

Run GitHub Label Sync on a repo ([passing in options](#configuration)):

```js
githubLabelSync({
	accessToken: 'xxxxxx',
	repo: 'myname/myrepo',
	labels: [
		// label config
	]
}).then((diff) => {
	console.log(diff);
});
```

The [available options](#configuration) are documented below.

When the promise resolves successfully, its value will be set to a diff between the labels found on GitHub, and the labels in your label config. The diff will look like this:

```js
[
	// This is a "missing" diff, it indicates that a label
	// present in your local config is not present on GitHub.
	{
		name: 'local-label-name',
		type: 'missing',
		actual: null,
		expected: {
			name: 'local-label-name',
			color: 'ff0000'
		}
	},
	// This is a "changed" diff, it indicates that a label
	// present on GitHub has diverged from your local config.
	// This could mean that either somebody has modified the
	// label manually on GitHub, or the local config has
	// been updated.
	{
		name: 'local-label-name',
		type: 'changed',
		actual: {
			name: 'remote-label-name',
			color: '00ff00'
		},
		expected: {
			name: 'local-label-name',
			color: 'ff0000'
		}
	},
	// This is an "added" diff, it indicates that a label
	// is present on GitHub but not in your local config.
	{
		name: 'remote-label-name',
		type: 'added',
		actual: {
			name: 'remote-label-name',
			color: 'ff0000'
		},
		expected: null
	}
]
```


Label Config File
----------

The labels to sync with are defined as an array in either JavaScript, JSON or YAML. The array must contain only label objects, which look like this:

As JSON:

```json
{
	"name": "mylabel",
	"color": "ff0000",
	"aliases": [],
	"description": "optional description"
}
```

As YAML:

```yaml
- name: mylabel
  color: "ff0000"
  aliases: []
  description: optional description
```

The `name` property refers to the label name and the `color` property should be set to the color of the label as a hex code without the leading `#`.

The `aliases` property is optional. When GitHub Label Sync is determining whether to update or delete/create a label it will use the aliases property to prevent used labels from being deleted.

For example, given the following config, GitHub Label Sync will look for labels on GitHub named either "feature" or "enhancement" then _update_ them to match the newer config rather than deleting them.

```json
{
	"name": "type: feature",
	"color": "00ff00",
	"aliases": [
		"enhancement",
		"feature"
	]
}
```

You can find a full example label configuration in this repository ([JSON](labels.json) / [YAML](labels.yml)).


Configuration
-------------

### `accessToken`

_String_. The [GitHub access token][access-tokens] to use when fetching/updating labels. This must be an access token that has permission to write to the repository you want to sync labels with.

```js
githubLabelSync({
	accessToken: 'xxxxxx'
});
```

On the command-line this can be set with either the `access-token` flag or the `GITHUB_ACCESS_TOKEN` environment variable:

```sh
github-label-sync --access-token xxxxxx
GITHUB_ACCESS_TOKEN=xxxxxx github-label-sync
```

### `allowAddedLabels`

_Boolean_. Whether to allow labels on GitHub which are not specified in your [label config](#label-config-file). If `true`, they are allowed and will be left alone. If `false`, they will be deleted. Default: `false`.

```js
githubLabelSync({
	allowAddedLabels: true
});
```

The command-line `allow-added-labels` flag corresponds to this option:

```
github-label-sync --allow-added-labels
```

### `dryRun`

_Boolean_. Whether to perform a dry run, only making safe "read" requests to the GitHub API. If `true`, label changes will not be executed on GitHub. If `false`, label changes will be executed. Default: `false`.

```js
githubLabelSync({
	dryRun: true
});
```

The command-line `dry-run` flag corresponds to this option:

```
github-label-sync --dry-run
```

### `labels`

_Array_. Your label configuration. See the section on [label config file](#label-config-file).

```js
githubLabelSync({
	labels: []
});
```

On the command-line this can be set with the `labels` flag which should point to a JSON or YAML file.

### `repo`

_String_. The GitHub repo to sync labels to. This should include the user and repo names, e.g. "Financial-Times/ft-origami".

```js
githubLabelSync({
	repo: 'Financial-Times/ft-origami'
});
```

The command-line accepts the repo as an argument after the options:

```
github-label-sync Financial-Times/ft-origami
```


Contributing
------------

To contribute to GitHub Label Sync, clone this repo locally and commit your code on a separate branch.

Please write unit tests for your code, and check that everything works by running the following before opening a pull-request:

```sh
npm test               # run the full test suite
npm run lint           # run the linter
npm run test-unit      # run the unit tests
npm run test-coverage  # run the unit tests with coverage reporting
```


License
-------

This software is published by the Financial Times under the [MIT licence][license].



[access-tokens]: https://github.com/settings/tokens
[license]: http://opensource.org/licenses/MIT
[node]: https://nodejs.org/
[npm]: https://www.npmjs.com/


CHANGELOG.md

# Changelog

## 2.0.0 (2020-06-03)

  * Drop support for node 4,5,6,7,8,9,10,11.

## 1.3.0 (2017-08-15)

  * Support URLs in the `--labels` flag

## 1.2.1 (2017-05-16)

  * Support repositories with a lot of labels

## 1.2.0 (2016-04-25)

  * Update the default label set
    * Add "status: good starter issue"
    * Add "type: breaking"
    * Add "type: discussion"
    * Add alias for "status: wontfix"

## 1.1.0 (2016-03-15)

  * Update the default label set
    * Add alias for "status: help wanted"

## 1.0.0 (2016-03-15)

  * Initial release


test/unit/lib/github-label-api.js
'use strict';

const assert = require('proclaim');
const mockery = require('mockery');

describe('lib/github-label-api', () => {
	let octonode;
	let githubLabelApi;

	beforeEach(() => {
		octonode = require('../mock/octonode');
		mockery.registerMock('octonode', octonode);
		githubLabelApi = require('../../../lib/github-label-api');
	});

	it('should export a function', () => {
		assert.isFunction(githubLabelApi);
	});

	describe('githubLabelApi(accessToken)', () => {
		const accessToken = 'mock-github-access-token';
		let instance;

		beforeEach(() => {
			instance = githubLabelApi(accessToken);
		});

		it('should create an octonode API client with `accessToken`', () => {
			assert.calledOnce(octonode.client);
			assert.calledWithExactly(octonode.client.firstCall, accessToken);
		});

		it('should return an object', () => {
			assert.isObject(instance);
		});

		describe('returned object', () => {

			it('should have an `apiClient` property', () => {
				assert.isDefined(instance.apiClient);
			});

			describe('.apiClient', () => {

				it('should be set to the created octonode API client', () => {
					assert.strictEqual(instance.apiClient, octonode.client.firstCall.returnValue);
				});

			});

			it('should have a `getLabels` method', () => {
				assert.isFunction(instance.getLabels);
			});

			describe('.getLabels(repo)', () => {
				const labels = ['foo', 'bar', 'baz'];
				const repo = 'foo/bar';
				let returnedPromise;

				beforeEach(() => {
					instance.apiClient.get.yieldsAsync(null, 200, labels);
					returnedPromise = instance.getLabels(repo);
				});

				it('should return a promise', () => {
					assert.instanceOf(returnedPromise, Promise);
				});

				it('should make a GET request to the repo labels endpoint', () => {
					assert.calledOnce(instance.apiClient.get);
					assert.calledWith(instance.apiClient.get, `/repos/${repo}/labels`, {page: 1, per_page: 100});
				});

				describe('.then()', () => {
					let resolvedValue;

					beforeEach((done) => {
						returnedPromise.then((value) => {
							resolvedValue = value;
							done();
						}).catch(done);
					});

					it('should resolve with the repo labels', () => {
						assert.deepEqual(resolvedValue, labels);
					});

				});

				describe('when the API client errors', () => {
					const apiClientError = new Error('API client error');

					beforeEach(() => {
						instance.apiClient.get.yieldsAsync(apiClientError);
						returnedPromise = instance.getLabels(repo);
					});

					describe('.catch()', () => {
						let caughtError;

						beforeEach((done) => {
							returnedPromise.then(done).catch((error) => {
								caughtError = error;
								done();
							});
						});

						it('should fail with the API client error', () => {
							assert.strictEqual(caughtError, apiClientError);
						});

						it('should augment the API client error with request information', () => {
							assert.strictEqual(caughtError.method, 'GET');
							assert.strictEqual(caughtError.endpoint, `/repos/${repo}/labels`);
						});

					});

				});

				describe('when the API response with a non-200 status', () => {

					beforeEach(() => {
						instance.apiClient.get.yieldsAsync(null, 404);
						returnedPromise = instance.getLabels(repo);
					});

					describe('.catch()', () => {
						let caughtError;

						beforeEach((done) => {
							returnedPromise.then(done).catch((error) => {
								caughtError = error;
								done();
							});
						});

						it('should fail with a status error', () => {
							assert.instanceOf(caughtError, Error);
							assert.strictEqual(caughtError.message, 'API responded with 404 status');
						});

					});

				});

			});

			it('should have a `createLabel` method', () => {
				assert.isFunction(instance.createLabel);
			});

			describe('.createLabel(repo, label)', () => {
				const createdLabel = {name: 'foo', color: 'ff0000', url: 'bar'};
				const label = {name: 'foo', color: 'ff0000'};
				const repo = 'foo/bar';
				let returnedPromise;

				beforeEach(() => {
					instance.apiClient.post.yieldsAsync(null, 201, createdLabel);
					returnedPromise = instance.createLabel(repo, label);
				});

				it('should return a promise', () => {
					assert.instanceOf(returnedPromise, Promise);
				});

				it('should make a POST request to the repo labels endpoint', () => {
					assert.calledOnce(instance.apiClient.post);
					assert.calledWith(instance.apiClient.post, `/repos/${repo}/labels`, label);
				});

				describe('.then()', () => {
					let resolvedValue;

					beforeEach((done) => {
						returnedPromise.then((value) => {
							resolvedValue = value;
							done();
						}).catch(done);
					});

					it('should resolve with the created label', () => {
						assert.strictEqual(resolvedValue, createdLabel);
					});

				});

				describe('when the API client errors', () => {
					const apiClientError = new Error('API client error');

					beforeEach(() => {
						instance.apiClient.post.yieldsAsync(apiClientError);
						returnedPromise = instance.createLabel(repo, label);
					});

					describe('.catch()', () => {
						let caughtError;

						beforeEach((done) => {
							returnedPromise.then(done).catch((error) => {
								caughtError = error;
								done();
							});
						});

						it('should fail with the API client error', () => {
							assert.strictEqual(caughtError, apiClientError);
						});

						it('should augment the API client error with request information', () => {
							assert.strictEqual(caughtError.method, 'POST');
							assert.strictEqual(caughtError.endpoint, `/repos/${repo}/labels`);
						});

					});

				});

				describe('when the API response with a non-201 status', () => {

					beforeEach(() => {
						instance.apiClient.post.yieldsAsync(null, 404);
						returnedPromise = instance.createLabel(repo, label);
					});

					describe('.catch()', () => {
						let caughtError;

						beforeEach((done) => {
							returnedPromise.then(done).catch((error) => {
								caughtError = error;
								done();
							});
						});

						it('should fail with a status error', () => {
							assert.instanceOf(caughtError, Error);
							assert.strictEqual(caughtError.message, 'API responded with 404 status');
						});

					});

				});

			});

			it('should have an `updateLabel` method', () => {
				assert.isFunction(instance.updateLabel);
			});

			describe('.updateLabel(repo, labelName, label)', () => {
				const label = {name: 'foo', color: 'ff0000'};
				const labelName = 'baz qux';
				const updatedLabel = {name: 'foo', color: 'ff0000', url: 'bar'};
				const repo = 'foo/bar';
				let returnedPromise;

				beforeEach(() => {
					instance.apiClient.patch.yieldsAsync(null, 200, updatedLabel);
					returnedPromise = instance.updateLabel(repo, labelName, label);
				});

				it('should return a promise', () => {
					assert.instanceOf(returnedPromise, Promise);
				});

				it('should make a PATCH request to the repo labels endpoint (encoding `labelName`)', () => {
					assert.calledOnce(instance.apiClient.patch);
					assert.calledWith(instance.apiClient.patch, `/repos/${repo}/labels/baz%20qux`, label);
				});

				describe('.then()', () => {
					let resolvedValue;

					beforeEach((done) => {
						returnedPromise.then((value) => {
							resolvedValue = value;
							done();
						}).catch(done);
					});

					it('should resolve with the updated label', () => {
						assert.strictEqual(resolvedValue, updatedLabel);
					});

				});

				describe('when the API client errors', () => {
					const apiClientError = new Error('API client error');

					beforeEach(() => {
						instance.apiClient.patch.yieldsAsync(apiClientError);
						returnedPromise = instance.updateLabel(repo, labelName, label);
					});

					describe('.catch()', () => {
						let caughtError;

						beforeEach((done) => {
							returnedPromise.then(done).catch((error) => {
								caughtError = error;
								done();
							});
						});

						it('should fail with the API client error', () => {
							assert.strictEqual(caughtError, apiClientError);
						});

						it('should augment the API client error with request information', () => {
							assert.strictEqual(caughtError.method, 'PATCH');
							assert.strictEqual(caughtError.endpoint, `/repos/${repo}/labels/baz%20qux`);
						});

					});

				});

				describe('when the API response with a non-200 status', () => {

					beforeEach(() => {
						instance.apiClient.patch.yieldsAsync(null, 404);
						returnedPromise = instance.updateLabel(repo, labelName, label);
					});

					describe('.catch()', () => {
						let caughtError;

						beforeEach((done) => {
							returnedPromise.then(done).catch((error) => {
								caughtError = error;
								done();
							});
						});

						it('should fail with a status error', () => {
							assert.instanceOf(caughtError, Error);
							assert.strictEqual(caughtError.message, 'API responded with 404 status');
						});

					});

				});

			});

			it('should have a `deleteLabel` method', () => {
				assert.isFunction(instance.deleteLabel);
			});

			describe('.deleteLabel(repo, labelName)', () => {
				const labelName = 'baz qux';
				const repo = 'foo/bar';
				let returnedPromise;

				beforeEach(() => {
					instance.apiClient.del.yieldsAsync(null, 204);
					returnedPromise = instance.deleteLabel(repo, labelName);
				});

				it('should return a promise', () => {
					assert.instanceOf(returnedPromise, Promise);
				});

				it('should make a DELETE request to the repo labels endpoint (encoding `labelName`)', () => {
					assert.calledOnce(instance.apiClient.del);
					assert.calledWith(instance.apiClient.del, `/repos/${repo}/labels/baz%20qux`, {});
				});

				describe('.then()', () => {
					let resolvedValue;

					beforeEach((done) => {
						returnedPromise.then((value) => {
							resolvedValue = value;
							done();
						}).catch(done);
					});

					it('should resolve with no value', () => {
						assert.isUndefined(resolvedValue);
					});

				});

				describe('when the API client errors', () => {
					const apiClientError = new Error('API client error');

					beforeEach(() => {
						instance.apiClient.del.yieldsAsync(apiClientError);
						returnedPromise = instance.deleteLabel(repo, labelName);
					});

					describe('.catch()', () => {
						let caughtError;

						beforeEach((done) => {
							returnedPromise.then(done).catch((error) => {
								caughtError = error;
								done();
							});
						});

						it('should fail with the API client error', () => {
							assert.strictEqual(caughtError, apiClientError);
						});

						it('should augment the API client error with request information', () => {
							assert.strictEqual(caughtError.method, 'DELETE');
							assert.strictEqual(caughtError.endpoint, `/repos/${repo}/labels/baz%20qux`);
						});

					});

				});

				describe('when the API response with a non-204 status', () => {

					beforeEach(() => {
						instance.apiClient.del.yieldsAsync(null, 404);
						returnedPromise = instance.deleteLabel(repo, labelName);
					});

					describe('.catch()', () => {
						let caughtError;

						beforeEach((done) => {
							returnedPromise.then(done).catch((error) => {
								caughtError = error;
								done();
							});
						});

						it('should fail with a status error', () => {
							assert.instanceOf(caughtError, Error);
							assert.strictEqual(caughtError.message, 'API responded with 404 status');
						});

					});

				});

			});

		});

	});

});


test/unit/lib/github-label-sync.js
'use strict';

const assert = require('proclaim');
const mockery = require('mockery');
const sinon = require('sinon');

describe('lib/github-label-sync', () => {
	let actionLabelDiff;
	let calculateLabelDiff;
	let extend;
	let githubLabelApi;
	let githubLabelSync;
	let log;
	let stringifyLabelDiff;

	beforeEach(() => {
		actionLabelDiff = require('../mock/action-label-diff');
		mockery.registerMock('./action-label-diff', actionLabelDiff);

		calculateLabelDiff = require('../mock/calculate-label-diff');
		mockery.registerMock('./calculate-label-diff', calculateLabelDiff);

		extend = sinon.spy(require('node.extend'));
		mockery.registerMock('node.extend', extend);

		githubLabelApi = require('../mock/github-label-api');
		mockery.registerMock('./github-label-api', githubLabelApi);

		log = require('../mock/log');

		stringifyLabelDiff = require('../mock/stringify-label-diff');
		mockery.registerMock('./stringify-label-diff', stringifyLabelDiff);

		githubLabelSync = require('../../../lib/github-label-sync');
	});

	it('should export a function', () => {
		assert.isFunction(githubLabelSync);
	});

	it('should have a `defaults` property', () => {
		assert.isObject(githubLabelSync.defaults);
	});

	describe('.defaults', () => {
		let defaults;

		beforeEach(() => {
			defaults = githubLabelSync.defaults;
		});

		it('should have an `accessToken` property', () => {
			assert.isNull(defaults.accessToken);
		});

		it('should have an `allowAddedLabels` property', () => {
			assert.isFalse(defaults.allowAddedLabels);
		});

		it('should have a `dryRun` property', () => {
			assert.isFalse(defaults.dryRun);
		});

		it('should have a `format` property', () => {
			assert.isObject(defaults.format);
		});

		it('should have a `format.diff` method', () => {
			assert.isFunction(defaults.format.diff);
		});

		it('should have a `format.success` method', () => {
			assert.isFunction(defaults.format.success);
		});

		it('should have a `format.warning` method', () => {
			assert.isFunction(defaults.format.warning);
		});

		it('should have a `labels` property', () => {
			assert.isArray(defaults.labels);
		});

		it('should have a `log` property', () => {
			assert.isObject(defaults.log);
		});

		it('should have a `log.info` method', () => {
			assert.isFunction(defaults.log.info);
		});

		it('should have a `log.warn` method', () => {
			assert.isFunction(defaults.log.warn);
		});

		it('should have a `repo` property', () => {
			assert.isNull(defaults.repo);
		});

	});

	describe('githubLabelSync(options)', () => {
		let apiClient;
		let labelsFromApi;
		let labelDiff;
		let labelDiffActions;
		let labelDiffStringified;
		let options;
		let returnedPromise;

		beforeEach(() => {
			apiClient = githubLabelApi.mockClient;

			// Mock API response
			labelsFromApi = [
				{
					name: 'foo',
					color: 'ff0000'
				}
			];
			apiClient.getLabels.resolves(labelsFromApi);

			// Mock label diff calculation
			labelDiff = [
				{
					name: 'foo',
					type: 'missing',
					actual: null,
					expected: {
						name: 'foo',
						color: 'ff0000'
					}
				}
			];
			calculateLabelDiff.returns(labelDiff);

			// Mock label diff actions
			labelDiffActions = [
				Promise.resolve()
			];
			actionLabelDiff.returns(labelDiffActions);

			// Mock label diff stringifying
			labelDiffStringified = [
				'foo stringified',
				'bar stringified'
			];
			stringifyLabelDiff.returns(labelDiffStringified);

			// Spy on Promise.all
			sinon.spy(Promise, 'all');

			// Mock options
			options = {
				accessToken: 'mock-github-access-token',
				labels: [
					{
						name: 'bar',
						color: '00ff00'
					}
				],
				log: log,
				repo: 'foo/bar'
			};
			returnedPromise = githubLabelSync(options);
		});

		afterEach(() => {
			Promise.all.restore();
		});

		it('should default the options', () => {
			assert.calledOnce(extend);
			assert.isTrue(extend.firstCall.args[0]);
			assert.isObject(extend.firstCall.args[1]);
			assert.strictEqual(extend.firstCall.args[2], githubLabelSync.defaults);
			assert.strictEqual(extend.firstCall.args[3], options);
		});

		it('should create a githubLabelApi client with `options.accessToken`', () => {
			assert.calledOnce(githubLabelApi);
			assert.calledWithExactly(githubLabelApi.firstCall, options.accessToken);
		});

		it('should return a promise', function() {
			assert.instanceOf(returnedPromise, Promise);
		});

		describe('.then()', () => {
			let resolvedValue;

			beforeEach((done) => {
				returnedPromise.then((value) => {
					resolvedValue = value;
					done();
				}).catch(done);
			});

			it('should get all of the labels for the repo in `options.repo`', () => {
				assert.calledOnce(apiClient.getLabels);
				assert.calledWithExactly(apiClient.getLabels, options.repo);
			});

			it('should diff the labels returned from the API against `options.labels`', () => {
				assert.calledOnce(calculateLabelDiff);
				assert.calledWithExactly(calculateLabelDiff, labelsFromApi, options.labels);
			});

			it('should stringify and log the returned diff', () => {
				assert.calledOnce(stringifyLabelDiff);
				assert.calledWith(stringifyLabelDiff, labelDiff);
				assert.calledWith(log.info, 'foo stringified');
				assert.calledWith(log.info, 'bar stringified');
			});

			it('should convert the returned diff to promises', () => {
				assert.calledOnce(actionLabelDiff);
				assert.isObject(actionLabelDiff.firstCall.args[0]);
				assert.strictEqual(actionLabelDiff.firstCall.args[0].apiClient, apiClient);
				assert.deepEqual(actionLabelDiff.firstCall.args[0].diff, labelDiff);
				assert.strictEqual(actionLabelDiff.firstCall.args[0].repo, options.repo);
			});

			it('should execute all of the diff promises', () => {
				assert.calledWith(Promise.all, labelDiffActions);
			});

			it('should log success', () => {
				assert.calledWith(log.info, 'Labels updated');
				assert.neverCalledWith(log.info, 'Labels are already up to date');
			});

			it('should resolve with the label diff', () => {
				assert.deepEqual(resolvedValue, labelDiff);
			});

		});

		describe('when no labels need to be updated', () => {

			beforeEach(() => {
				log.info.reset();
				calculateLabelDiff.returns([]);
				actionLabelDiff.returns([]);
				returnedPromise = githubLabelSync(options);
			});

			describe('.then()', () => {

				beforeEach((done) => {
					returnedPromise.then(() => {
						done();
					}).catch(done);
				});

				it('should log that labels were up to date', () => {
					assert.calledWith(log.info, 'Labels are already up to date');
					assert.neverCalledWith(log.info, 'Labels updated');
				});

			});

		});

		describe('when the `dryRun` option is `true`', () => {

			beforeEach(() => {
				options.dryRun = true;
				actionLabelDiff.reset();
				Promise.all.reset();
				returnedPromise = githubLabelSync(options);
			});

			describe('.then()', () => {

				beforeEach((done) => {
					returnedPromise.then(() => {
						done();
					}).catch(done);
				});

				it('should not convert the returned diff to promises', () => {
					assert.notCalled(actionLabelDiff);
				});

				it('should not execute any diff promises', () => {
					assert.neverCalledWith(Promise.all, labelDiffActions);
				});

			});

		});

		describe('when the `allowAddedLabels` option is `true`', () => {

			beforeEach(() => {
				labelDiff.push({
					name: 'bar',
					type: 'added',
					actual: {
						name: 'bar',
						color: '00ff00'
					},
					expected: null
				});
				options.allowAddedLabels = true;
				actionLabelDiff.reset();
				stringifyLabelDiff.reset();
				returnedPromise = githubLabelSync(options);
			});

			describe('.then()', () => {

				beforeEach((done) => {
					returnedPromise.then(() => {
						done();
					}).catch(done);
				});

				it('should not include "added" diffs in stringification', () => {
					assert.calledOnce(stringifyLabelDiff);
					assert.deepEqual(stringifyLabelDiff.firstCall.args[0], [labelDiff[0]]);
				});

				it('should not convert "added" diffs to promises', () => {
					assert.calledOnce(actionLabelDiff);
					assert.isObject(actionLabelDiff.firstCall.args[0]);
					assert.deepEqual(actionLabelDiff.firstCall.args[0].diff, [labelDiff[0]]);
				});

			});

		});

	});

});


lib/calculate-label-diff.js
'use strict';

module.exports = calculateLabelDiff;

function calculateLabelDiff(currentLabels, configuredLabels) {
	const diff = [];
	const resolvedLabels = [];
	configuredLabels.forEach((configuredLabel) => {

		// Get current labels which match the configured label
		const matches = currentLabels.filter((currentLabel) => {
			if (currentLabel.name.toLowerCase() === configuredLabel.name.toLowerCase()) {
				return true;
			}
			if (configuredLabel.aliases && configuredLabel.aliases.map(label => label.toLowerCase()).indexOf(currentLabel.name.toLowerCase()) !== -1) {
				return true;
			}
		});

		// If we have no matches, the configured label is missing
		if (matches.length === 0) {
			return diff.push(createMissingEntry(configuredLabel));
		}

		// Always take the first match
		const matchedLabel = matches[0];
		resolvedLabels.push(matchedLabel);

		const matchedDescription = getLabelDescription(matchedLabel);
		const configuredDescription = getLabelDescription(configuredLabel, matchedDescription);

		// If we have a match, but properties are not equal
		if (configuredLabel.name !== matchedLabel.name ||
			configuredLabel.color !== matchedLabel.color ||
			configuredDescription !== matchedDescription
		) {
			return diff.push(createChangedEntry(matchedLabel, configuredLabel));
		}

	});
	currentLabels.filter(label => resolvedLabels.indexOf(label) === -1).forEach((currentLabel) => {
		diff.push(createAddedEntry(currentLabel));
	});
	return diff;
}

function getLabelDescription(label, fallback = '') {
	if (label.description === undefined) {
		return fallback;
	}
	return (label.description && label.description.trim()) || '';
}

function createMissingEntry(expectedLabel) {
	const missingEntry = {
		name: expectedLabel.name,
		type: 'missing',
		actual: null,
		expected: {
			name: expectedLabel.name,
			color: expectedLabel.color
		}
	};
	const expectedDescription = getLabelDescription(expectedLabel);
	if (expectedDescription) {
		missingEntry.expected.description = expectedDescription;
	}
	return missingEntry;
}

function createChangedEntry(actualLabel, expectedLabel) {
	const changedEntry = {
		name: actualLabel.name,
		type: 'changed',
		actual: {
			name: actualLabel.name,
			color: actualLabel.color
		},
		expected: {
			name: expectedLabel.name,
			color: expectedLabel.color
		}
	};

	const actualDescription = getLabelDescription(actualLabel);
	const expectedDescription = getLabelDescription(expectedLabel, actualDescription);
	if (actualDescription === expectedDescription && !actualDescription) {
		return changedEntry;
	}

	changedEntry.actual.description = actualDescription;
	changedEntry.expected.description = expectedDescription;
	return changedEntry;
}

function createAddedEntry(actualLabel) {
	const addedEntry = {
		name: actualLabel.name,
		type: 'added',
		actual: {
			name: actualLabel.name,
			color: actualLabel.color
		},
		expected: null
	};
	const actualDescription = getLabelDescription(actualLabel);
	if (actualDescription) {
		addedEntry.actual.description = actualDescription;
	}
	return addedEntry;
}


lib/github-label-api.js
'use strict';

const github = require('octonode');

module.exports = createApiClient;

class ApiClient {

	constructor (accessToken) {
		this.apiClient = github.client(accessToken);
	}

	getLabels (repo) {
		return new Promise((resolve, reject) => {
			const allLabels = [];
			const endpoint = `/repos/${repo}/labels`;
			const params = {
				page: 1,
				per_page: 100
			};
			const getCallback = (error, status, labels) => {
				if (error) {
					error.method = 'GET';
					error.endpoint = endpoint;
					return reject(error);
				}
				if (status !== 200) {
					return reject(new Error(`API responded with ${status} status`));
				}
				allLabels.push.apply(allLabels, labels);
				if (labels.length === params.per_page) {
					params.page += 1;
					this.apiClient.get(endpoint, params, getCallback);
				} else {
					resolve(allLabels);
				}
			};
			this.apiClient.get(endpoint, params, getCallback);
		});
	}

	createLabel (repo, label) {
		return new Promise((resolve, reject) => {
			const endpoint = `/repos/${repo}/labels`;
			this.apiClient.post(endpoint, label, (error, status, createdLabel) => {
				if (error) {
					error.method = 'POST';
					error.endpoint = endpoint;
					return reject(error);
				}
				if (status !== 201) {
					return reject(new Error(`API responded with ${status} status`));
				}
				resolve(createdLabel);
			});
		});
	}

	updateLabel (repo, labelName, label) {
		labelName = encodeURIComponent(labelName);
		return new Promise((resolve, reject) => {
			const endpoint = `/repos/${repo}/labels/${labelName}`;
			this.apiClient.patch(endpoint, label, (error, status, updatedLabel) => {
				if (error) {
					error.method = 'PATCH';
					error.endpoint = endpoint;
					return reject(error);
				}
				if (status !== 200) {
					return reject(new Error(`API responded with ${status} status`));
				}
				resolve(updatedLabel);
			});
		});
	}

	deleteLabel (repo, labelName) {
		labelName = encodeURIComponent(labelName);
		return new Promise((resolve, reject) => {
			const endpoint = `/repos/${repo}/labels/${labelName}`;
			this.apiClient.del(endpoint, {}, (error, status) => {
				if (error) {
					error.method = 'DELETE';
					error.endpoint = endpoint;
					return reject(error);
				}
				if (status !== 204) {
					return reject(new Error(`API responded with ${status} status`));
				}
				resolve();
			});
		});
	}

}

function createApiClient(accessToken) {
	return new ApiClient(accessToken);
}
